### PR TITLE
Use std::optional instead of llvm::Optional

### PR DESCRIPTION
**Pull request by jayfoad**

### CODE REVIEW

**Review by dstutt — APPROVED** [review]
LGTM - presumably any version that doesn't work with std::optional is so far back that it doesn't matter? (Based on the fact that there were previously some flag day changes that you've removed)

**Comment by jayfoad** [comment]
> presumably any version that doesn't work with std::optional is so far back that it doesn't matter?

That's what I'm hoping, unless anyone is deliberately building with a very old version of LLVM. I'm hoping some CI will fail if that is the case.

**Comment by jasilvanus** [comment]
Note that `std::optional<>::operator*` is not required to check for a value, and libstdc++'s implementation does not do any checking. This differs from `llvm::Optional<>::operator*` which asserted in debug builds.

In contrast, `std::optional<>::value()` does check and throws/aborts if there is no value.

Not sure how we should handle this, it seems in LLVM it was not cared much about and just replaced anyways.
At least, we should be aware and maybe use more `value()` instead of `operator*` depending on context.

**Comment by jayfoad** [comment]
It sounds to me like `std::optional<>::operator*` has the behaviour we want for Release builds. For Debug builds this is surely just a QoI issue of the C++ library. I see that libcxx has assertions in its implementation of `operator*`.

**Comment by jayfoad** [comment]
And in libstdc++-12 I see things like this:
```
      // The _M_get operations have _M_engaged as a precondition.
      constexpr _Tp&
      _M_get() noexcept
      {
        __glibcxx_assert(this->_M_is_engaged());
        return static_cast<_Dp*>(this)->_M_payload._M_get();
      }
```
where `operator*` calls `_M_get`.

**Comment by jasilvanus** [comment]
It appears one can set `_GLIBCXX_ASSERTIONS` for debug builds to get these asserts. Maybe I should start doing that :)

Still, one should be aware of this difference -- one can no longer just run a debug build and rely on `std::optional` catching missing value errors.

**Comment by amdvlk-admin** [comment]
8dde1e1050170801894693ba1f09444a95a6b732 Jenkins build error.
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/disassembler/Disassembler.cpp:184:59: error: conversion from ‘llvm::Optional<llvm::StringRef>’ to non-scalar type ‘std::optional<llvm::StringRef>’ requested
   std::optional<StringRef> mcpu = m_objFile->tryGetCPUName();
                                   ~~~~~~~~~~~~~~~~~~~~~~~~^~
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/disassembler/Disassembler.cpp:189:67: error: no match for ‘operator<<’ (operand types are ‘llvm::raw_ostream’ and ‘std::optional<llvm::StringRef>’)
   m_ostream <<  // llvm-mc -triple=  << m_tripleName <<   -mcpu=  << mcpu <<   n ;
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~
In file included from /jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llvm-project/llvm/include/llvm/Support/Error.h:27:0,
                 from /jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llvm-project/llvm/include/llvm/BinaryFormat/MsgPackReader.h:36,
                 from /jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llvm-project/llvm/include/llvm/BinaryFormat/MsgPackDocument.h:20,
--
/usr/include/c++/7/ostream:682:5: error: no type named ‘type’ in ‘struct std::enable_if<false, void>’
/usr/include/c++/7/ostream:574:5: note: candidate: template<class _Traits> std::basic_ostream<char, _Traits>& std::operator<<(std::basic_ostream<char, _Traits>&, const unsigned char*)
     operator<<(basic_ostream<char, _Traits>& __out, const unsigned char* __s)
     ^~~~~~~~
/usr/include/c++/7/ostream:574:5: note:   template argument deduction/substitution failed:
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/disassembler/Disassembler.cpp:189:70: note:   ‘llvm::raw_ostream’ is not derived from ‘std::basic_ostream<char, _Traits>’

**Comment by amdvlk-admin** [comment]
### Test summary for commit 598b40ad143d5f0a9ca8ed413b9d6166baf767b3
<details><summary>CTS tests (Failed: 0/282101)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 38864/70526 (55.1%)</li>
<li>Failed: 0/70526 (0.0%)</li>
<li>Not Supported: 31662/70526 (44.9%)</li><li>Warnings: 0/70526 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi2x</summary><ul>

<li>Passed: 39114/70526 (55.5%)</li>
<li>Failed: 0/70526 (0.0%)</li>
<li>Not Supported: 31412/70526 (44.5%)</li><li>Warnings: 0/70526 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35959/70528 (51.0%)</li>
<li>Failed: 0/70528 (0.0%)</li>
<li>Not Supported: 34569/70528 (49.0%)</li><li>Warnings: 0/70528 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 35861/70521 (50.9%)</li>
<li>Failed: 0/70521 (0.0%)</li>
<li>Not Supported: 34660/70521 (49.1%)</li><li>Warnings: 0/70521 (0.0%)</li>
</ul></details>


</ul></details>

File: llpc/context/llpcCompiler.h [hunk 1]
```diff
@@ -85,8 +85,8 @@ class GraphicsShaderCacheChecker {
 private:
   Compiler *m_compiler;
   Context *m_context;
-  llvm::Optional<CacheAccessor> m_nonFragmentCacheAccessor;
-  llvm::Optional<CacheAccessor> m_fragmentCacheAccessor;
+  std::optional<CacheAccessor> m_nonFragmentCacheAccessor;
```
Comment (amdrexu) [inline llpc/context/llpcCompiler.h:88]: Note that `#include <optional>` is missing in this header while in the internal porting this is added.

File: lgc/patch/PatchEntryPointMutate.cpp [hunk 2]
```diff
@@ -1491,7 +1491,7 @@ void PatchEntryPointMutate::determineUnspilledUserDataArgs(ArrayRef<UserDataArg>
                                                            IRBuilder<> &builder,
                                                            SmallVectorImpl<UserDataArg> &unspilledArgs) {
 
-  Optional<UserDataArg> spillTableArg;
+  std::optional<UserDataArg> spillTableArg;
```
Comment (amdrexu) [inline lgc/patch/PatchEntryPointMutate.cpp:1494]: `#include <optional>` is missing

File: llpc/context/llpcCompiler.h (hunk 1 above)
Comment (jayfoad) [reply llpc/context/llpcCompiler.h:88]: Yes, sorry, I added that #include when I was testing the internal version.

Comment (amdrexu) [reply llpc/context/llpcCompiler.h:88]: Never mind. Just need a tiny fix :)

**Pull request merged**